Formula/git-filter-repo.rb

```diff
@@ -0,0 +1,17 @@
+class GitFilterRepo < Formula
+  desc "Quickly rewrite git repository history"
+  homepage "https://github.com/newren/git-filter-repo"
+  url "https://github.com/newren/git-filter-repo/releases/download/v2.24.0/git-filter-repo-2.24.0.tar.xz"
+  sha256 "92188d3c44b9ff0dd40dfeed72859e0a088f775c12fb24c4e3e27a8064cfcc84"
+
+  depends_on "python"
+
+  def install
+    bin.install "git-filter-repo"
+    man1.install "Documentation/man1/git-filter-repo.1"
+  end
+
+  test do
+    system "#{bin}/git-filter-repo", "--version"
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
@SMillerDev there's a test suite in the project itself:
- https://github.com/newren/git-filter-repo/blob/v2.24.0/Makefile#L14-L15

however, I can't find how to execute it in the `test` stage, because it needs the worktree (unpacked sources, not installed files).

I've seen some other formulas doing `make test` in `install` stage,
but this shouldn't be done here, for several reasons:
1. upstream runs the test suite by themselves
2. it's slow

```
$ time brew reinstall git-filter-repo -s --verbose --debug
...
==> Reinstalling git-filter-repo
...
==> make test
cd t && time ./run_coverage
ok 1 - check: basic -> basic-filename using '--path filename'
ok 2 - check: basic -> basic-twenty using '--path twenty'
ok 3 - check: basic -> basic-ten using '--path ten'
...
# passed all 11 test(s)
1..11
./run_coverage: line 27: python3-coverage: command not found
./run_coverage: line 28: python3-coverage: command not found
./run_coverage: line 29: python3-coverage: command not found

real	1m5.314s
user	0m29.181s
sys	0m31.401s
==> Cleaning
==> Fixing /usr/local/opt/git-filter-repo/bin/git-filter-repo permissions from 755 to 555
==> Finishing up
ln -s ../Cellar/git-filter-repo/2.24.0/bin/git-filter-repo git-filter-repo
ln -s ../../../Cellar/git-filter-repo/2.24.0/share/man/man1/git-filter-repo.1 git-filter-repo.1
/usr/bin/sandbox-exec -f /private/tmp/homebrew20191209-40746-1f1ow7f.sb nice ruby -W0 -I $LOAD_PATH -- /usr/local/Homebrew/Library/Homebrew/postinstall.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/git-filter-repo.rb -s --verbose --debug
/usr/local/Homebrew/Library/Homebrew/postinstall.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/git-filter-repo.rb
==> Summary
🍺  /usr/local/Cellar/git-filter-repo/2.24.0: 6 files, 235.8KB, built in 1 minute 9 seconds

real	1m11.472s
user	0m33.448s
sys	0m33.175s
$
```

Reply:
@newren: can you suggest some trivial self-test that brew maintainers require?

> We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

sorry for bringing you into this discussion, opening ticket in https://github.com/newren/git-filter-repo seemed even worse idea.

Reply:
EDIT: or rather asking for a review, maybe I missed something trivial.

commit: 57bc166319

output:

```
➔ brew test git-filter-repo --verbose --debug
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/git-filter-repo.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/python.rb
Testing git-filter-repo
/usr/bin/sandbox-exec -f /private/tmp/homebrew20191209-43064-1lu0mgb.sb ruby -W0 -I $LOAD_PATH -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/git-filter-repo.rb --verbose --debug
/usr/local/Homebrew/Library/Homebrew/test.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/git-filter-repo.rb
==> /usr/local/Cellar/git-filter-repo/2.24.0/bin/git-filter-repo --version
f3e8e0f8a87c
==> git init
Initialized empty Git repository in /private/tmp/git-filter-repo-test-20191209-43065-94ooc1/.git/
==> git config user.name BrewTestBot
==> git config user.email BrewTestBot@example.com
==> git add foo
==> git commit -m foo
[master (root-commit) 366b6ae] foo
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 foo
==> /usr/local/Cellar/git-filter-repo/2.24.0/bin/git-filter-repo --path-rename=foo:bar --force
Parsed 1 commits
HEAD is now at 7f39085 foo

New history written in 0.10 seconds; now repacking/cleaning...
Repacking your repo and cleaning out old unneeded objects
Completely finished after 0.30 seconds.
```

Reply:
This isn't about unit/code tests though, this should test if the homebrew install succeeded. The last snippet you send seems good, it tests something the user would normally run (as opposed to a version check).